test/conftest.py

```diff
@@ -2300,43 +2301,72 @@ def _copy_files(self, req):
     async def AttemptStart(self, stream):
         request: api_pb2.AttemptStartRequest = await stream.recv_message()
         fn_definition = self.app_functions.get(request.function_id)
+
+        self.fcidx += 1
+        function_call_id = f"fc-{self.fcidx}"
+        input_id = f"in-{self.n_inputs}"
+        self.n_inputs += 1
+        self.add_function_call_input(function_call_id, request.input, input_id, 0)
+
         retry_policy = fn_definition.retry_policy if fn_definition else None
         # TODO(ryan): implement attempt token logic
         await stream.send_message(
             api_pb2.AttemptStartResponse(attempt_token="bogus_attempt_token", retry_policy=retry_policy)
         )
 
     async def AttemptAwait(self, stream):
-        # TODO(ryan): Eventually we may want to invoke the user's function and return a result.
-        # For now we just return a dummy response
+        # TODO(dxia): implement attempt token logic
 
-        # To test client retries, tests can configure outputs to fail some number of times.
+        # To test client retries for internal failures, tests can configure outputs to fail some number of times.
         status = api_pb2.GenericResult.GENERIC_STATUS_SUCCESS
         if self.attempt_await_failures_remaining > 0:
             status = api_pb2.GenericResult.GENERIC_STATUS_INTERNAL_FAILURE
             self.attempt_await_failures_remaining = self.attempt_await_failures_remaining - 1
 
+        if self.function_call_inputs:
+            function_call_id = f"fc-{self.fcidx}"
+            (idx, input_id, retry_count), (args, kwargs) = self.function_call_inputs.pop(function_call_id)[0]
+            self.fcidx -= 1
+            try:
+                res = self._function_body(*args, **kwargs)
+            except Exception as e:
+                res = e
+                status = api_pb2.GenericResult.GENERIC_STATUS_FAILURE
+        else:
```

Comment:
now that we're invoking the user's function, can we get rid of this default behavior?

Reply:
oh, I see in the implementation of FunctionGetOutputs that if there is no input to process, it blocks waiting for one to appear. I guess that's not a trivial thing to implement, so maybe we can tackle that in another PR.